File: Lernwerk/Services/Storage/PresentationStore.swift
import Foundation
import UIKit

/// Presentations as one JSON file each, their pictures in a shared media folder (same layout as on Android).
@MainActor
final class PresentationStore: ObservableObject {
    @Published private(set) var presentations: [Presentation] = []

    private let root: URL
    private let mediaDirectory: URL
    private var pendingWrites: [String: Task<Void, Never>] = [:]
    private let images = NSCache<NSString, UIImage>()

    init(root: URL? = nil) {
        let base = root ?? FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0].appendingPathComponent("Presentations", isDirectory: true)
        self.root = base
        mediaDirectory = base.appendingPathComponent("media", isDirectory: true)
        try? FileManager.default.createDirectory(at: mediaDirectory, withIntermediateDirectories: true)
        let decoder = JSONDecoder()
        let files = (try? FileManager.default.contentsOfDirectory(at: base, includingPropertiesForKeys: nil)) ?? []
        presentations = files
            .filter { $0.pathExtension == "json" }
            .compactMap { try? decoder.decode(Presentation.self, from: Data(contentsOf: $0)) }
            .sorted { $0.updatedAt > $1.updatedAt }
    }

    func presentation(_ id: String) -> Presentation? {
        presentations.first { $0.id == id }
    }

    func add(_ presentation: Presentation) {
        presentations.insert(presentation, at: 0)
        write(presentation, immediately: true)
    }

    /// Keeps the list current right away and writes the file shortly after, so dragging does not hit the disk.
    func update(_ presentation: Presentation) {
        guard let index = presentations.firstIndex(where: { $0.id == presentation.id }) else { return }
        var updated = presentation
        updated.updatedAt = Int64(Date().timeIntervalSince1970 * 1000)
        presentations[index] = updated
        write(updated, immediately: false)
    }

    func delete(_ presentation: Presentation) {
        presentations.removeAll { $0.id == presentation.id }
        pendingWrites.removeValue(forKey: presentation.id)?.cancel()
        try? FileManager.default.removeItem(at: file(presentation.id))
    }

    func duplicate(_ presentation: Presentation) {
        var copy = presentation
        copy.id = UUID().uuidString
        copy.title = "\(presentation.title) (Kopie)"
        copy.createdAt = Int64(Date().timeIntervalSince1970 * 1000)
        copy.updatedAt = copy.createdAt
        add(copy)
    }

    private func file(_ id: String) -> URL {
        root.appendingPathComponent("\(id).json")
    }

    private func write(_ presentation: Presentation, immediately: Bool) {
        pendingWrites.removeValue(forKey: presentation.id)?.cancel()
        let url = file(presentation.id)
        pendingWrites[presentation.id] = Task {
            if !immediately {
                try? await Task.sleep(nanoseconds: 600_000_000)
            }
            guard !Task.isCancelled, let data = try? JSONEncoder().encode(presentation) else { return }
            try? data.write(to: url, options: .atomic)
        }
    }

    // Media

    func mediaURL(_ name: String) -> URL {
        mediaDirectory.appendingPathComponent(name)
    }

    /// Stores a picture and returns its file name for an image element.
    func saveMedia(_ data: Data, fileExtension: String) throws -> String {
        let name = "\(UUID().uuidString).\(fileExtension)"
        try data.write(to: mediaURL(name), options: .atomic)
        return name
    }

    /// Stores a picture under a name chosen by the caller, e.g. from an imported file.
    func saveMedia(_ data: Data, named name: String) throws {
        try data.write(to: mediaURL(name), options: .atomic)
    }

    func mediaData(_ name: String) -> Data? {
        try? Data(contentsOf: mediaURL(name))
    }

    func image(_ name: String) -> UIImage? {
        if let cached = images.object(forKey: name as NSString) { return cached }
        guard let image = UIImage(contentsOfFile: mediaURL(name).path) else { return nil }
        images.setObject(image, forKey: name as NSString)
        return image
    }

    /// All pictures a set of slides uses, for drawing.
    func images(for slides: [Slide]) -> [String: UIImage] {
        var result: [String: UIImage] = [:]
        for slide in slides {
            for element in slide.elements {
                if let name = element.image, result[name] == nil, let image = image(name) { result[name] = image }
            }
        }
        return result
    }
}

File: android/app/src/main/java/de/maxifrz/lernwerk/data/PresentationStore.kt
package de.maxifrz.lernwerk.data

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.LruCache
import androidx.compose.runtime.mutableStateListOf
import de.maxifrz.lernwerk.present.Presentation
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import java.io.File
import java.util.UUID

/** Presentations as one JSON file each, their pictures in a shared media folder. */
class PresentationStore(context: Context) {
    private val root = File(context.filesDir, "presentations").apply { mkdirs() }
    private val mediaDir = File(root, "media").apply { mkdirs() }
    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
        encodeDefaults = true
    }
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val writeLock = Mutex()
    private val pendingWrites = mutableMapOf<String, Job>()
    private val bitmaps = object : LruCache<String, Bitmap>(64 * 1024 * 1024) {
        override fun sizeOf(key: String, value: Bitmap) = value.byteCount
    }

    val presentations = mutableStateListOf<Presentation>()

    init {
        root.listFiles { file -> file.extension == "json" }
            ?.mapNotNull { file -> runCatching { json.decodeFromString(Presentation.serializer(), file.readText()) }.getOrNull() }
            ?.sortedByDescending { it.updatedAt }
            ?.let { presentations += it }
    }

    fun presentation(id: String) = presentations.firstOrNull { it.id == id }

    fun add(presentation: Presentation) {
        presentations.add(0, presentation)
        write(presentation, immediately = true)
    }

    /** Keeps the list current right away and writes the file shortly after, so dragging does not hit the disk. */
    fun update(presentation: Presentation) {
        val updated = presentation.copy(updatedAt = System.currentTimeMillis())
        val index = presentations.indexOfFirst { it.id == updated.id }
        if (index < 0) return
        presentations[index] = updated
        write(updated, immediately = false)
    }

    fun delete(presentation: Presentation) {
        presentations.removeAll { it.id == presentation.id }
        pendingWrites.remove(presentation.id)?.cancel()
        scope.launch { writeLock.withLock { file(presentation.id).delete() } }
    }

    private fun file(id: String) = File(root, "$id.json")

    private fun write(presentation: Presentation, immediately: Boolean) {
        pendingWrites.remove(presentation.id)?.cancel()
        pendingWrites[presentation.id] = scope.launch {
            if (!immediately) delay(600)
            val text = json.encodeToString(Presentation.serializer(), presentation)
            writeLock.withLock {
                val target = file(presentation.id)
                val temp = File(root, "${presentation.id}.json.tmp")
                temp.writeText(text)
                if (!temp.renameTo(target)) {
                    target.writeText(text)
                    temp.delete()
                }
            }
        }
    }

    // Media

    fun mediaFile(name: String) = File(mediaDir, name)

    /** Stores a picture and returns its file name for an image element. */
    suspend fun saveMedia(bytes: ByteArray, extension: String): String = withContext(Dispatchers.IO) {
        val name = "${UUID.randomUUID()}.$extension"
        mediaFile(name).writeBytes(bytes)
        name
    }

    fun mediaBytes(name: String): ByteArray? = mediaFile(name).takeIf { it.exists() }?.readBytes()

    /** A picture for drawing, scaled down to at most [maxSize] pixels on its longer side and cached. */
    suspend fun bitmap(name: String, maxSize: Int = 1600): Bitmap? {
        bitmaps.get("$name@$maxSize")?.let { return it }
        return withContext(Dispatchers.IO) {
            val file = mediaFile(name)
            if (!file.exists()) return@withContext null
            val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
            BitmapFactory.decodeFile(file.path, bounds)
            var sample = 1
            while (maxOf(bounds.outWidth, bounds.outHeight) / (sample * 2) >= maxSize) sample *= 2
            BitmapFactory.decodeFile(file.path, BitmapFactory.Options().apply { inSampleSize = sample })
                ?.also { bitmaps.put("$name@$maxSize", it) }
        }
    }

    /** Size of a stored picture in pixels, for giving new image elements the right aspect ratio. */
    suspend fun imageSize(name: String): Pair<Int, Int>? = withContext(Dispatchers.IO) {
        val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeFile(mediaFile(name).path, bounds)
        if (bounds.outWidth > 0) bounds.outWidth to bounds.outHeight else null
    }
}
